### ISSUE TITLE
Excel add-in keeps disappearing from Ribbon

### ISSUE THREAD

**Issue by jimbrew** [Excel-DNA/ExcelDna#557]
Hello,

What version of Excel is supported by ExcalDna v1.6?

I have an add-in that keeps being removed from the ribbon. Excel will take a while to load then it is removed from the ribbon. The addin is not being disabled as it is still present in file > options > addins.

It is also still ticked under the Excel Addins window. In order to enable a user has to untick, close Excel, open Excel, add it back in.

**Comment by govert** [comment]
Excel 2007 and later is supported by Excel-DNA.
It sounds like your add-in is failing during the load. Can you reproduce the problem under a debugger?
Two things are happening during the load:
1. Some of your code is JIT-compiled as Excel-DNA checks for features to export, which means dependencies are being resolved and additional assemblies loaded.
2. Code like the initializer of you ExcelRibbon class and the GetCustomUI is running. If you have anything complicated here, you might want to take a closer look.

**Comment by govert** [comment]
Are you sure it's not doing some extra work when loading the second time, for example trying load settings or cached information that it stores the first time?
If you're still stuck and would like some direct support, you are welcome to contact me.

**Comment by jimbrew** [comment]
I am not 100% what goes on behind the scenes. Unfortunately, the person who coded the add-in no longer works at the company and I know very little about it. 

It did work on previous versions of Excel and now we have few clients saying it keeps getting disabled.

I updated the nugets, reduced the dependencies that aren't needed. I am at a bit of a loss with this and apologies I cannot explain the issues further.

**Comment by andrewkittredge** [comment]
You might look at the .dna XML file.  It seems like the most recent version of Excel is stricter about its correctness.  For instance, we had an ```onLoad``` pointing to a function deleted that was ignored until the most recent Excel release.

**Comment by jimbrew** [comment]
Thanks for your help so far - I created a hello world excel add-in with ribbon and all works fine. So, it will be an issue with the code or load times that is causing excel to disable my add-in.

I do get this error though
![image](https://user-images.githubusercontent.com/35928790/211783713-3fbce9de-d63b-4802-a409-204f020fe041.png)

**Comment by jimbrew** [comment]
Hi just an update.

I have managed to figure out what is causing the issues. Excel does not close down properly after the add-in is used which is why it only happens on next load. There is still a hanging instance of Excel.

If I use the add-in close Excel, quit any running instances using task manager and reopen excel the add-in is present.

Any ideas how to make sure my application closes properly? Do I need anything in the `AutoClose `method?

```
    public class DNAAddin : IExcelAddIn
    {
        private static readonly Logger Logger = LogManager.GetCurrentClassLogger();

        public void AutoOpen()
        {
            try
            {
                Logger.Info("Binding to Macro");

                ExcelRegistration
                    .GetExcelFunctions()
                    .ProcessParamsRegistrations()
                    .RegisterFunctions();
            }
            catch(Exception ex)
            {
                Logger.Error(ex);
            }
        }

        public void AutoClose()
        {
            try
            {
                Logger.Info("Auto closing");
            }
            catch(Exception ex)
            {
                Logger.Error(ex);
            }
        }
    }
```

**Comment by govert** [comment]
The problem is usually that you are using the Excel COM object model from another (not the main) thread.
All access to the Excel COM objects should be on the main thread, not a Task, or a timer callback that might run on another thread. Code that is running on another thread can schedule code to run on the main thread (where the COM work can be done) with a call to `ExcelAsyncUtil.QueueAsMacro`.

**Comment by jimbrew** [comment]
Yes, looking at the code there is a method that displays a window under a different thread.

I will check [this](https://docs.excel-dna.net/performing-asynchronous-work/) out and convert `new Thread` to something more safe.
